pkg/exchange/ftx/stream.go

```diff
@@ -102,15 +117,69 @@ func (s *Stream) SetPublicOnly() {
 	atomic.StoreInt32(&s.publicOnly, 1)
 }
 
-func (s *Stream) Subscribe(channel types.Channel, symbol string, _ types.SubscribeOptions) {
-	if channel != types.BookChannel {
+func (s *Stream) Subscribe(channel types.Channel, symbol string, option types.SubscribeOptions) {
+	if channel == types.BookChannel {
+		s.addSubscription(websocketRequest{
+			Operation: subscribe,
+			Channel:   orderBookChannel,
+			Market:    toLocalSymbol(TrimUpperString(symbol)),
+		})
+
+	} else if channel == types.KLineChannel {
+		// FTX does not support kline channel, do polling
+		go s.subscribeKLine(symbol, option)
+	} else {
 		panic("only support book channel now")
 	}
-	s.addSubscription(websocketRequest{
-		Operation: subscribe,
-		Channel:   orderBookChannel,
-		Market:    toLocalSymbol(TrimUpperString(symbol)),
-	})
+}
+
+func (s *Stream) handleChannelKlineMessage() {
+	for {
+		kline := <-s.klineMessage
+
+		if kline.Closed {
+			s.EmitKLineClosed(kline)
+			continue
+		}
+
+		s.EmitKLine(kline)
+	}
+}
+
+func (s *Stream) subscribeKLine(symbol string, option types.SubscribeOptions) {
+	interval := types.Interval(option.Interval)
+	if !isIntervalSupportedInKLine(interval) {
+		logger.Errorf("not supported kline interval %s", option.Interval)
+		return
+	}
+
+	for {
+		if !s.isConnected {
+			time.Sleep(time.Second * 2)
+			continue
+		}
+
+		// get the last kline
+		since := time.Now().Add(time.Duration(-1*(interval.Minutes())) * time.Minute)
+		kline, err := s.exchange.QueryKLines(s.ctx, symbol, interval, types.KLineQueryOptions{
+			StartTime: &since,
+		})
+		if err != nil {
```

Comment:
wrap this with retry?

Reply:
I found pkg/util/retry.go but can not find anywhere used it. Should we use this or use https://pkg.go.dev/gopkg.in/retry.v1 ?